src/managedwriter/stream_connection.ts

```diff
@@ -106,62 +122,53 @@ export class StreamConnection extends EventEmitter {
 
   private handleError = (err: gax.GoogleError) => {
     this.trace('on error', err, JSON.stringify(err));
-    if (this.shouldReconnect(err)) {
-      this.reconnect();
-      return;
-    }
-    let nextPendingWrite = this.getNextPendingWrite();
-    if (this.isPermanentError(err)) {
-      this.trace('found permanent error', err);
-      while (nextPendingWrite) {
-        this.ackNextPendingWrite(err);
-        nextPendingWrite = this.getNextPendingWrite();
-      }
-      this.emit('error', err);
-      return;
-    }
-    if (this.isRequestError(err) && nextPendingWrite) {
+    this._lastConnectionError = err;
+    const nextPendingWrite = this.getNextPendingWrite();
+    if (nextPendingWrite) {
       this.trace(
         'found request error with pending write',
         err,
         nextPendingWrite
       );
-      this.ackNextPendingWrite(err);
+      this.handleRetry(err);
+    }
+    if (this.isRetryableError(err) && this.listenerCount('error') === 0) {
       return;
     }
     this.emit('error', err);
   };
 
-  private shouldReconnect(err: gax.GoogleError): boolean {
-    const reconnectionErrorCodes = [
-      gax.Status.UNAVAILABLE,
-      gax.Status.RESOURCE_EXHAUSTED,
+  private handleRetry(err: gax.GoogleError) {
```

Comment:
two questions
1) does this handle timeout logic at all? 
2) Would you consider modifying gax instead of creating more custom retry logic?

Reply:
@leahecole:
1. when a timeout happen, what is the status code on gax ? I think it is `ABORTED` right ? which is this case is retryable.
2. Yes we considered, but there is custom retry logic here that is tightly coupled with the BQ Storage service, like. triggering a reconnect or not to the service. Other service libraries would definitely be benefited by adding support for retries on BiDi streaming, but we would still have to add some logic here that is specific for BQ Storage.

Reply:
Timeout is `DEADLINE_EXCEEDED`! ⏰ Re #2, sounds good for now